drv/fpga-api/src/lib.rs

```diff
@@ -293,14 +307,51 @@ pub fn await_fpga_ready(
 /// Load a bitstream.
 pub fn load_bitstream(
     fpga: &mut Fpga,
-    data: &[u8],
+    auxflash: &mut AuxFlash,
+    blob: AuxFlashBlob,
     bitstream_type: BitstreamType,
-    chunk_len: usize,
+    expected_checksum: [u8; 32],
 ) -> Result<(), FpgaError> {
-    let mut bitstream = fpga.start_bitstream_load(bitstream_type)?;
+    // Do an initial read to verify the checksum
+    let mut scratch_buf = [0u8; 128];
+    let mut pos = blob.start;
+    let mut sha = Sha3_256::new();
+    while pos != blob.end {
+        let amount = (blob.end - pos).min(scratch_buf.len() as u32);
+        let chunk = &mut scratch_buf[0..(amount as usize)];
+
+        auxflash
+            .read_slot_with_offset(blob.slot, pos, chunk)
+            .map_err(|_| FpgaError::AuxReadError)?;
+        pos += amount;
+
+        sha.update(chunk);
+    }
+    let sha_out: [u8; 32] = sha.finalize().into();
+    if sha_out != expected_checksum {
+        return Err(FpgaError::AuxChecksumMismatch);
+    }
 
-    for chunk in data.chunks(chunk_len) {
-        bitstream.continue_load(chunk)?;
+    // Then, stream the data into the FPGA, recalculating the checksum to
+    // protect against TOCTOU attacks.
+    let mut bitstream = fpga.start_bitstream_load(bitstream_type)?;
+    let mut pos = blob.start;
+    let mut sha = Sha3_256::new();
+    while pos != blob.end {
+        let amount = (blob.end - pos).min(scratch_buf.len() as u32);
+        let chunk = &mut scratch_buf[0..(amount as usize)];
+
+        auxflash
+            .read_slot_with_offset(blob.slot, pos, chunk)
+            .map_err(|_| FpgaError::AuxReadError)?;
+        pos += amount;
+
+        sha.update(&chunk);
+        bitstream.continue_load(&chunk)?;
+    }
```

Comment:
Would it be slightly more maintainable to extract the "read and hash the slot" into a separate function, since this is almost identical to the above? Would need some kind of callback to handle the bitstream load. Untested, but something like

```rust
fn read_and_hash_slot(
    auxflash: &mut AuxFlash,
    blob: &AuxFlashBlob,
    scratch_buf: &mut [u8],
    callback: &mut dyn FnMut(&[u8]) -> Result<(), FpgaError>,
) -> Result<[u8; 32], FpgaError> {
    let mut pos = blob.start;
    let mut sha = Sha3_256::new();
    while pos < blob.end {
        let amount = (blob.end - pos).min(scratch_buf.len() as u32);
        let chunk = &mut scratch_buf[0..(amount as usize)];

        auxflash
            .read_slot_with_offset(blob.slot, pos, chunk)
            .map_err(|_| FpgaError::AuxReadError)?;
        pos += amount;

        sha.update(&chunk);
        callback(&chunk)?;
    }
    Ok(sha.finalize.into())
}
```

which would change most of this function to something like

```rust
    let mut scratch_buf = [0u8; 128];

    // Do an initial read to verify the checksum
    let sha_out = read_and_hash_slot(auxflash, &blob, &mut scratch_buf, &mut |_| Ok(()))?;
    if sha_out != expected_checksum {
        return Err(FpgaError::AuxChecksumMismatch);
    }

    // Then, stream the data into the FPGA, recalculating the checksum to
    // protect against TOCTOU attacks.
    let mut bitstream = fpga.start_bitstream_load(bitstream_type)?;
    let sha_out = read_and_hash_slot(auxflash, &blob, &mut scratch_buf,
        &mut |chunk| bitstream.continue_load(chunk))?;
    if sha_out != expected_checksum {
        return Err(FpgaError::AuxChecksumMismatch);
    }
```

Reply:
Yeah, this is great!